server/chat-plugins/net-filters.ts

```diff
@@ -0,0 +1,244 @@
+/**
+ * Neural net chatfilters.
+ * These are in a separate file so that they don't crash the other filters.
+ * (issues with globals, etc)
+ * by Mia.
+ * @author mia-pi-git
+ */
+
+import {QueryProcessManager, QueryProcessWrapper} from '../../lib/process-manager';
+import {FS} from '../../lib/fs';
+import {Utils} from '../../lib/utils';
+import {Config} from '../config-loader';
+import {Repl} from '../../lib/repl';
+
+const PATH = "config/chat-plugins/net.json";
+const NUM_PROCESSES = Config.netfilterprocesses || 1;
+const PM_TIMEOUT = 2 * 60 * 60 * 1000; // training can be _really_ slow
+const WHITELIST = ["mia"];
+
+interface NetQuery {
+	data: string | TrainingLine[];
+	type: "run" | "train" | "save" | "load";
+	options?: AnyObject;
+}
+
+interface TrainingLine {
+	input: string;
+	// "|flag" (bad) "|ok" (good)
+	output: string;
+}
+
+function modelExists() {
+	try {
+		require.resolve('brain.js');
+	} catch (e) {
+		return false;
+	}
+	return true;
+}
+
+export class NeuralNetChecker {
+	model: import('brain.js').recurrent.LSTM | null;
+	constructor(path?: string) {
+		try {
+			this.model = new (require('brain.js').recurrent.LSTM)();
+		} catch (e) {
+			this.model = null;
+		}
+		if (path) this.load(path);
+	}
+	async train(data: TrainingLine[], iterations?: number) {
+		// 100 has good perf but is still effective
+		if (!iterations) iterations = 100;
+		const now = Date.now();
+		await FS(PATH).copyFile(PATH + '.backup');
+		if (!this.model) throw new Error(`Attempting to train with no model installed`);
+		this.model.train(data, {iterations});
+		this.save();
+		return Date.now() - now; // time data is helpful for training
+	}
+	save(path = PATH) {
+		if (!this.model) return {};
+		const state = this.model.toJSON();
+		FS(path).writeUpdate(() => JSON.stringify(state));
+		return state;
+	}
+	load(path: string) {
+		if (!FS(path).existsSync()) return;
+		const data = JSON.parse(FS(path).readSync());
+		this.model?.fromJSON(data);
+	}
+	run(data: string) {
+		let result = '';
+		if (!this.model) return result;
+		try {
+			result = this.model.run(data);
+		} catch (e) {}
+		// usually means someone didn't train it, carry on
+		// acceptable to drop since training is very slow
+		return result;
+	}
+	static async train(data: TrainingLine[]) {
+		// do the training in a separate process (which we never add to PM.processes so it doesn't get pinged for other requests)
+		const result = await PM.queryTemporaryProcess({type: 'train', data});
+		// load it into the main process that we're querying
+		await PM.query({type: 'load', data: PATH});
```

Comment:
Wait, this only works if you have exactly one process... PM isn't designed to be used like that. :s

Reply:
I guess it's fine but `Config.netfilterprocesses` can't exist if you do this.

Reply:
Alright. We probably can do with that for now? Or I could just iterate each process and query each of them to load them?